apps/server/src/config/repositoryConfig.ts

```diff
@@ -0,0 +1,113 @@
+
+const definitionBaseRepositorySchema = z.object({
+	BASE_URL: z.string().url(),
+	CODE: z.string(),
+	NAME: z.string(),
+	PAGINATION_SIZE: z.coerce.number().optional(),
+	INDEX_NAME: z
+		.string()
+		.refine((value) => !/[<" \\/,|>?*]/.test(value), {
+			// criteria: https://www.elastic.co/guide/en/elasticsearch/reference/7.17/indices-create-index.html
+			message: `INDEX_NAME cannot include characters '<', '"', space character, '\\' , '/', ',', '|', '>', '?', or '*'`,
+		})
+		.refine((value) => !/^[-_+]/.test(value), {
+			message: 'INDEX_NAME cannot start with -, _, or +',
+		})
+		.refine((value) => value !== '.' && value !== '..', {
+			message: 'INDEX_NAME cannot be . or ..',
+		})
+		.refine((value) => /^[^A-Z]*$/.test(value), {
+			message: 'INDEX_NAME must be lowercase and contain only alphanumeric characters',
+		})
+		.refine((value) => Buffer.byteLength(value, 'utf-8') <= 255, {
+			message: 'INDEX_NAME cannot be longer than 255 bytes',
+		}),
+});
+
+const definitionLyricRepositorySchema = z.object({
+	TYPE: z.literal(repositoryTypes.Values.LYRIC),
+	LYRIC_VALID_DATA_ONLY: z.coerce.boolean().default(true),
+	LYRIC_CATEGORY_ID: z.coerce.number(),
+});
+export const lyricSchemaDefinition = definitionBaseRepositorySchema.and(definitionLyricRepositorySchema);
+const isLyricRepository = (data: unknown): data is z.infer<typeof lyricSchemaDefinition> => {
+	return lyricSchemaDefinition.safeParse(data).success;
+};
+
+const definitionSongRepositorySchema = z.object({
+	TYPE: z.literal(repositoryTypes.Values.SONG),
+	SONG_INDEXABLE_STUDY_STATES: z.string().default('PUBLISHED'),
+	SONG_ANALYSIS_CENTRIC_ENABLED: z.coerce.boolean().default(true),
+	SONG_ORGANIZATION: z.string().optional(),
+	SONG_COUNTRY: z.string().optional(),
+});
+export const songSchemaDefinition = definitionBaseRepositorySchema.and(definitionSongRepositorySchema);
+const isSongRepository = (data: unknown): data is z.infer<typeof songSchemaDefinition> => {
+	return songSchemaDefinition.safeParse(data).success;
+};
+
+// Define the number of repositories based on the environment variables present
+const getRepoCount = (): number => {
+	let count = 0;
+	while (process.env[`MAESTRO_REPOSITORIES_${count}_BASE_URL`]) {
+		count++;
+	}
+	return count;
+};
+
+/**
+ * Search through the passed environment variables for any related to repositories and perform validation
+ * @param env Environment variables
+ * @returns an array of valid repositories
+ * @throws an error if repository is invalid
+ */
+export const validateRepositories = (env: NodeJS.ProcessEnv) => {
+	const repositorySchema = definitionBaseRepositorySchema.and(
+		z.discriminatedUnion('TYPE', [definitionLyricRepositorySchema, definitionSongRepositorySchema]),
+	);
+
+	const resultParsedRepositories: (z.infer<typeof lyricSchemaDefinition> | z.infer<typeof songSchemaDefinition>)[] = [];
+
+	const repoCount = getRepoCount();
+
+	// Loop through the repositories found
+	for (let i = 0; i < repoCount; i++) {
```

Comment:
please replace this with `repoCount.forEach` to make things more declarative.
#ImperativeSucks

Reply:
you're right... and in my defence, coffee hadn't kicked in.
my argument against imperative/procedural code stands, we're meant to tell the system "what" to do, not "how"